<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
<meta name="generator" content="pdoc3 0.11.0">
<title>epi_ml.utils.clean_hdf5 API documentation</title>
<meta name="description" content="Module: clean_hdf5 …">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/sanitize.min.css" integrity="sha512-y1dtMcuvtTMJc1yPgEqF0ZjQbhnc/bFhyvIyVNb9Zk5mIGtqVaAB1Ttl28su8AvFMOY0EwRbAe+HCLqj6W7/KA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/typography.min.css" integrity="sha512-Y1DYSb995BAfxobCkKepB1BqJJTPrOp3zPL74AWFugHHmmdcvO+C48WLrUOlhGMc0QG7AE3f7gmvvcrmX2fDoA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:1.5em;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:2em 0 .50em 0}h3{font-size:1.4em;margin:1.6em 0 .7em 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .2s ease-in-out}a:visited{color:#503}a:hover{color:#b62}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900;font-weight:bold}pre code{font-size:.8em;line-height:1.4em;padding:1em;display:block}code{background:#f3f3f3;font-family:"DejaVu Sans Mono",monospace;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em 1em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul ul{padding-left:1em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js" integrity="sha512-D9gUyxqja7hBtkWpPWGt9wfbfaMGVt9gnyCvYa+jojwwPHLCzUm5i8rpk7vD7wNee9bA35eYIjobYPaQuKS1MQ==" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => {
hljs.configure({languages: ['bash', 'css', 'diff', 'graphql', 'ini', 'javascript', 'json', 'plaintext', 'python', 'python-repl', 'rust', 'shell', 'sql', 'typescript', 'xml', 'yaml']});
hljs.highlightAll();
})</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>epi_ml.utils.clean_hdf5</code></h1>
</header>
<section id="section-intro">
<p>Module: clean_hdf5</p>
<p>This module provides functionality for cleaning and processing genomic data stored in HDF5 files. Specifically, it allows for applying blacklist filters onto HDF5 datasets to modify certain genomic intervals. It reads BED (Browser Extensible Data) files containing these intervals, preprocesses them, determines the positions that are subject to treatment based on the blacklist, and applies changes to HDF5 datasets.</p>
<p>The module primarily consists of the following functions:</p>
<pre><code>load_bed(path: Path | str) -&gt; BEDIntervals:
Loads the contents of a BED file.

preprocess_bed(bed: BEDIntervals) -&gt; Dict[str, BEDIntervals]:
Preprocesses the BED intervals.

get_positions_to_treat(blacklist_chrom_intervals: Dict[str, BEDIntervals], bin_resolution: int) -&gt; Dict[str, List[int]]:
Generates the positions to treat based on the bin resolution and blacklist intervals.

process_file(og_hdf5_path: Path, positions_to_treat: Dict[str, List[int]], output_dir: Path):
Cleans one HDF5 file and saves a copy with blacklisted regions set to zero. This function assumes the positions to treat contain chromosome vector indices for the right resolution.
</code></pre>
<p>main() -&gt; None:
The main driver function that organizes the entire process of reading, filtering and writing the HDF5 files.</p>
<p>This module also defines the following type aliases:</p>
<pre><code>BEDInterval = Tuple[str, int, int]: Represents a single genomic interval from a BED file.
BEDIntervals = List[BEDInterval]: Represents a list of genomic intervals from a BED file.
</code></pre>
<p>Dependencies:
This module requires the h5py library for HDF5 file operations, and the epi_ml package for argument parsing and directory checking utilities.</p>
<p>Usage:
Invoke the main function to process a list of HDF5 files according to command-line arguments:</p>
<pre><code>hdf5_list: A file containing a list of HDF5 filenames.
bed_filter: A path to a BED file containing the blacklist positions.
output_dir: The directory where the modified HDF5 files will be saved.
n_jobs (optional): The number of parallel jobs to run.
</code></pre>
<p>Example:
$ python clean_hdf5.py hdf5_list.txt blacklist.bed output_directory -n 4</p>
<p>Author:
Joanny Raby</p>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="epi_ml.utils.clean_hdf5.check_file"><code class="name flex">
<span>def <span class="ident">check_file</span></span>(<span>og_hdf5_path: Path, positions_to_treat: Dict[str, List[int]]) ‑> bool</span>
</code></dt>
<dd>
<div class="desc"><p>Checks one hdf5 file to ensure all designated positions are zero.</p>
<p>Supposes the positions to treat contain chromosome vector indices for the right resolution.</p>
<p>Returns: True if all positions are zero, False otherwise.</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.get_positions_to_treat"><code class="name flex">
<span>def <span class="ident">get_positions_to_treat</span></span>(<span>blacklist_chrom_intervals: Dict[str, BEDIntervals], bin_resolution: int) ‑> Dict[str, List[int]]</span>
</code></dt>
<dd>
<div class="desc"><p>Generate the positions to treat based on the bin resolution and blacklist intervals.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>blacklist_chrom_intervals</code></strong></dt>
<dd>A dictionary mapping chromosomes to sorted intervals.</dd>
<dt><strong><code>bin_resolution</code></strong></dt>
<dd>The bin resolution.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>A dictionary mapping chromosomes to a list of positions to be treated.</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.load_bed"><code class="name flex">
<span>def <span class="ident">load_bed</span></span>(<span>path: Path | str)</span>
</code></dt>
<dd>
<div class="desc"><p>Loads the contents of a BED file.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>path</code></strong></dt>
<dd>The path to the BED file.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>A list of tuples representing the BED intervals (chromosome, start, end).</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.main"><code class="name flex">
<span>def <span class="ident">main</span></span>(<span>) ‑> None</span>
</code></dt>
<dd>
<div class="desc"><p>Main function.</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.parse_arguments"><code class="name flex">
<span>def <span class="ident">parse_arguments</span></span>(<span>) ‑> argparse.Namespace</span>
</code></dt>
<dd>
<div class="desc"><p>argument parser for command line</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.preprocess_bed"><code class="name flex">
<span>def <span class="ident">preprocess_bed</span></span>(<span>bed: BEDIntervals) ‑> Dict[str, List[Tuple[str, int, int]]]</span>
</code></dt>
<dd>
<div class="desc"><p>Preprocesses the BED intervals.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>bed</code></strong></dt>
<dd>A list of tuples representing the BED intervals (chromosome, start, end).</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>A dictionary mapping chromosomes to sorted intervals.</p></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.print_h5py_structure"><code class="name flex">
<span>def <span class="ident">print_h5py_structure</span></span>(<span>file: str) ‑> None</span>
</code></dt>
<dd>
<div class="desc"><p>Prints the structure of an h5py file.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>file</code></strong></dt>
<dd>The path to the h5py file.</dd>
</dl></div>
</dd>
<dt id="epi_ml.utils.clean_hdf5.process_file"><code class="name flex">
<span>def <span class="ident">process_file</span></span>(<span>og_hdf5_path: Path, positions_to_treat: Dict[str, List[int]], output_dir: Path)</span>
</code></dt>
<dd>
<div class="desc"><p>Clean one hdf5 file. Save copy with regions that touch blacklisted regions to 0.</p>
<p>Supposes the positions to treat contain chromosome vector indices for the right resolution.</p></div>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="epi_ml.utils" href="index.html">epi_ml.utils</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="epi_ml.utils.clean_hdf5.check_file" href="#epi_ml.utils.clean_hdf5.check_file">check_file</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.get_positions_to_treat" href="#epi_ml.utils.clean_hdf5.get_positions_to_treat">get_positions_to_treat</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.load_bed" href="#epi_ml.utils.clean_hdf5.load_bed">load_bed</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.main" href="#epi_ml.utils.clean_hdf5.main">main</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.parse_arguments" href="#epi_ml.utils.clean_hdf5.parse_arguments">parse_arguments</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.preprocess_bed" href="#epi_ml.utils.clean_hdf5.preprocess_bed">preprocess_bed</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.print_h5py_structure" href="#epi_ml.utils.clean_hdf5.print_h5py_structure">print_h5py_structure</a></code></li>
<li><code><a title="epi_ml.utils.clean_hdf5.process_file" href="#epi_ml.utils.clean_hdf5.process_file">process_file</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.11.0</a>.</p>
</footer>
</body>
</html>
